src/wallet3/wallet.cpp

```diff
@@ -45,9 +49,26 @@ namespace wallet
     if (not tx_constructor)
       this->tx_constructor = std::make_shared<TransactionConstructor>(db, daemon_comms);
 
+    config.omq_rpc.sockname = file_path_from_default_datadir(config, config.omq_rpc.sockname).string();
     omq_server.set_omq(this->omq, config.omq_rpc);
 
     db->create_schema();
+    if (keyring)
+    {
+      keys = keyring;
+      db->save_keys(
+        tools::type_to_hex(keys->spend_private_key),
+        tools::type_to_hex(keys->spend_public_key),
+        tools::type_to_hex(keys->view_private_key),
+        tools::type_to_hex(keys->view_public_key)
+        );
```

Comment:
Doing the `save_keys` this way lets the WalletDB be dumb (i.e. doesn't know about KeyRing) but it just exported the knowledge to the caller of WalletDB which know has to know *both* about the KeyRing *and* which keys WalletDB cares about.

I think this would be a good case to use an interface for the keys instead, so that we can pass `db->save_keys(*keys)` rather than Wallet having to care about *which* keys it needs to pass: instead WalletDB can use the interface to grab out the keys it cares about.

If we use an interface this gets cleaned up: WalletDB only has to know that it is given something that provides the thing it cares about, WalletKeys has to provide those things, and Wallet doesn't have to care *at all* about which values it pulls out of KeyRing to shove into WalletDB.

The interface header could look like this:
```C++
// walletkeys.h
class WalletKeys {
public:
    virtual const crypto::secret_key& spend_privkey() const = 0;  
    virtual const crypto::public_key& spend_pubkey() const = 0;  
    virtual const crypto::secret_key& view_privkey() const = 0;  
    virtual const crypto::public_key& view_pubkey() const = 0;  
};  
```

and then WalletKeyring would inherit from this and expose these keys via implementations of the methods:
```C++
class WalletKeyring : public WalletKeys {  
    crypto::secret_key _ssk, _vsk;  
    crypto::public_key _spk, _vpk;  
    // ...  

public:
    const crypto::secret_key& spend_privkey() const override { return ssk; }  
    const crypto::public_key& spend_pubkey() const override { return spk; }  
    const crypto::secret_key& view_privkey() const override { return vsk; }  
    const crypto::public_key& view_pubkey() const override { return vpk; }  
};  
```

and `save_keys`  becomes:
```C++
void WalletDB::save_keys(const WalletKeys& keys) {
    // store keys->spend_privkey(), etc.
}

// from Wallet:
db->save_keys(*keyring);
```


Now WalletDB doesn't need to know anything about WalletKeyring *except* that it has the functions it cares about to get the keys it cares about out of it.

If, in the future, there are some other keys that need to be stored similarly (for example, a lokinet persistent key, or a X25519 key for SN access) it would just get added to the interface without having to touch

Reply:
For `load_keys` it could work similarly:
```C++
// walletdb:
struct DBKeys : public WalletKeys {
    crypto::secret_key ssk, vsk;
    crypto::public_key spk, vpk;

    const crypto::secret_key& spend_privkey() const override { return ssk; }
    const crypto::public_key& spend_pubkey() const override { return spk; }
    const crypto::secret_key& view_privkey() const override { return vsk; }
    const crypto::public_key& view_pubkey() const override { return vpk; }
};

DBKeys WalletDB::load_keys() {
    // make and return a filled-out DBKeys
};
```
and then WalletKeyring constructor would now take a single `const WalletKeys&` instead of 4 separate key values.